### PR TITLE
Problem: memiavl rootmulti integration is not thread-safe to use

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #977

Solution:
- add mutex to `memiavl.DB`.

The assumptions to concurrency:
- The methods on `DB` are protected by a mutex
- each call of `LoadVersion` loads a separate instance, in query scenarios, it should be immutable, the cache stores will handle the temporary writes.
- the `DB` for the state machine will handle writes through the `Commit` call, this method is the sole entry point for tree modifications, and there's no concurrency internally (the background snapshot rewrite is handled separately), so we don't need locks in the `Tree`.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db.go [hunk 1]
```diff
@@ -302,9 +342,68 @@ func (db *DB) RewriteSnapshotBackground() error {
...
+	return db.MultiTree.Version()
+}
+
+// LastCommitInfo returns the last commit info.
+func (db *DB) LastCommitInfo() *storetypes.CommitInfo {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	return db.MultiTree.LastCommitInfo()
+}
```
Comment (mmsqe) [inline memiavl/db.go:381]: do we need make a lock wrapper like?
```suggestion
func withLock[T any](db *DB, f func(*MultiTree) T) T {
	db.mtx.Lock()
	defer db.mtx.Unlock()
	return f(&db.MultiTree)
}

// LastCommitInfo wraps MultiTree LastCommitInfo to add a lock.
func (db *DB) LastCommitInfo() *storetypes.CommitInfo {
	return withLock(db, func(mtree *MultiTree) *storetypes.CommitInfo {
		return mtree.LastCommitInfo()
	})
}
```

Comment (yihuang) [reply memiavl/db.go:381]: the non-callback version seems simple enough?

Comment (mmsqe) [reply memiavl/db.go:381]: yea

**Review by mmsqe — APPROVED** [review]

File: memiavl/db.go [hunk 2]
```diff
@@ -39,6 +40,15 @@ type DB struct {
 	// invariant: the LastIndex always match the current version of MultiTree
 	wal             *wal.Log
 	pendingUpgrades []*TreeNameUpgrade
+
+	// The assumptions to concurrency:
+	// - The methods on DB are protected by a mutex
+	// - Aach call of LoadVersion loads a separate instance, in query scenarios,
```
Comment (mmsqe) [inline memiavl/db.go:46]:
```suggestion
	// - Each call of LoadVersion loads a separate instance, in query scenarios,
```

**Pull request merged**